### ISSUE TITLE
interference from docker-proxy in DNS resolution

### ISSUE THREAD

**Issue by rade** [weaveworks/weave#1192]
observed while investigating #1171...

I am seeing some `getent` requests stalling. Here's a tcpdump taken with `docker-ns weave tcpdump -w /tmp/dump1.pcap -ni eth0 port 53` and covering some normal requests, then the stalling ones, and then again some normal requests:
````
20:25:33.940327 IP 172.17.42.1.52028 > 172.17.2.88.53: 50166+ AAAA? bar.weave.local. (33)
20:25:33.940708 IP 172.17.2.88.53 > 172.17.42.1.52028: 50166 NXDomain 0/0/0 (33)
20:25:33.941572 IP 172.17.42.1.50230 > 172.17.2.88.53: 41134+ AAAA? bar.weave.local.weave.local. (45)
20:25:33.941839 IP 172.17.2.88.53 > 172.17.42.1.50230: 41134 NXDomain 0/0/0 (45)
20:25:33.942839 IP 172.17.42.1.48263 > 172.17.2.88.53: 38208+ A? bar.weave.local. (33)
20:25:33.943099 IP 172.17.2.88.53 > 172.17.42.1.48263: 38208* 2/0/0 A 10.2.4.2, A 10.2.4.130 (95)
20:25:34.004325 IP 172.17.42.1.50420 > 172.17.2.88.53: 10475+ PTR? 2.4.2.10.in-addr.arpa. (39)
20:25:34.004788 IP 172.17.2.88.53 > 172.17.42.1.50420: 10475* 1/0/0 PTR bar.weave.local. (89)
20:25:34.067281 IP 172.17.42.1.57400 > 172.17.2.88.53: 11215+ PTR? 130.4.2.10.in-addr.arpa. (41)
20:25:34.067986 IP 172.17.2.88.53 > 172.17.42.1.57400: 11215* 1/0/0 PTR bar.weave.local. (93)
20:25:34.476152 IP 172.17.2.91.49377 > 172.17.42.1.53: 9074+ AAAA? foo.weave.local. (33)
20:25:39.481341 IP 172.17.2.91.49377 > 172.17.42.1.53: 9074+ AAAA? foo.weave.local. (33)
20:25:44.486791 IP 172.17.42.1.60746 > 172.17.2.88.53: 10538+ AAAA? foo.weave.local.weave.local. (45)
20:25:44.488074 IP 172.17.2.88.53 > 172.17.42.1.60746: 10538 NXDomain 0/0/0 (45)
````

Notice how the normal requests have the docker bridge IP as the source, whereas the failing ones originate from a container IP and have the docker bridge as the destination IP.

Looking at the iptable packet counters...
````
root@host1:~# iptables -vnL -t nat | grep "dpt:53"
    0     0 MASQUERADE  tcp  --  *      *       172.17.2.102         172.17.2.102         tcp dpt:53
    0     0 MASQUERADE  udp  --  *      *       172.17.2.102         172.17.2.102         udp dpt:53
    0     0 DNAT       tcp  --  !docker0 *       0.0.0.0/0            172.17.42.1          tcp dpt:53 to:172.17.2.102:53
    0     0 DNAT       udp  --  !docker0 *       0.0.0.0/0            172.17.42.1          udp dpt:53 to:172.17.2.102:53
root@host1:~# iptables -vnL | grep "dpt:53"
    0     0 ACCEPT     tcp  --  !docker0 docker0  0.0.0.0/0            172.17.2.102         tcp dpt:53
    0     0 ACCEPT     udp  --  !docker0 docker0  0.0.0.0/0            172.17.2.102         udp dpt:53
````
shows that none of rules fire!

Disabling the docker-proxy by adding `--userland-proxy=false` to the `ExecStart` line in `/lib/systemd/system/docker.service`, and then bouncing docker with

````
systemctl daemon-reload 
systemctl restart docker
systemctl enable docker
````

makes the stalling problem disappear, and the packet counters now show some action

````
root@host1:~# iptables -vnL -t nat | grep "dpt:53"
    0     0 MASQUERADE  tcp  --  *      *       172.17.0.97          172.17.0.97          tcp dpt:53
    0     0 MASQUERADE  udp  --  *      *       172.17.0.97          172.17.0.97          udp dpt:53
    0     0 DNAT       tcp  --  *      *       0.0.0.0/0            172.17.42.1          tcp dpt:53 to:172.17.0.97:53
   36  2396 DNAT       udp  --  *      *       0.0.0.0/0            172.17.42.1          udp dpt:53 to:172.17.0.97:53
````

[Apparently](https://github.com/docker/docker/issues/8356) the docker-proxy is only used for localhost traffic, to work around some limitations of older kernels. Which would explain why we see problems in our DNS resolution - since that is local - but not for weave peer connectivity.

**Comment by squaremo** [comment]
Some context I've picked up from docker issues: The ability to switch the proxy off landed in Docker 1.7; it's not been removed entirely because Docker still supports RHEL 6 with its elderly kernel, but that support will end with Docker 1.8 (https://github.com/docker/docker/issues/14365).

We may not be the only people to see this behaviour; to wit, https://github.com/docker/docker/issues/7540. The expression of the problem is the other way around, but it seems to be the same inconsistency.

**Comment by rade** [comment]
https://github.com/docker/docker/issues/7540#issuecomment-94316789 is interesting.

**Comment by rade** [comment]
The following smoke test reproduces the problem for me:
````
#! /bin/bash

. ./config.sh

start_suite "stress restart"

weave_on $HOST1 launch

proxy docker_on $HOST1 run -dti --name=foo $SMALL_IMAGE /bin/sh
start_container_with_dns $HOST1 --name=baz

for i in $(seq 50); do
    proxy docker_on $HOST1 kill foo 1>/dev/null
    assert_no_dns_record $HOST1 baz foo.weave.local
    proxy docker_on $HOST1 start foo 1>/dev/null
done

end_suite
````
Running this with `DEBUG=1` shows the stalls.

I've taken packet captures in the host with `tcpdump -w /tmp/dump-docker0.pcap -ni docker0`. These show up some weirdness...

Here are the interfaces and their addresses that we have on the host:
````
docker0   Link encap:Ethernet  HWaddr 06:c6:e2:10:9a:be  
          inet addr:172.17.42.1  Bcast:0.0.0.0  Mask:255.255.0.0
eth0      Link encap:Ethernet  HWaddr 08:00:27:9b:9f:1d  
          inet addr:10.0.2.15  Bcast:10.0.2.255  Mask:255.255.255.0
eth1      Link encap:Ethernet  HWaddr 08:00:27:e2:51:20  
          inet addr:192.168.48.11  Bcast:192.168.48.255  Mask:255.255.255.0
veth5479e8c Link encap:Ethernet  HWaddr 32:fc:9e:71:ac:8c  
veth78819fd Link encap:Ethernet  HWaddr 06:c6:e2:10:9a:be 
vethbafcb12 Link encap:Ethernet  HWaddr b2:38:14:4d:ac:c8  
vethwepl2434 Link encap:Ethernet  HWaddr ce:dc:93:7f:51:2d  
vethwepl3004 Link encap:Ethernet  HWaddr 22:01:b5:1f:f6:c2  
vethwepl7588 Link encap:Ethernet  HWaddr ee:f5:ca:75:a9:bf  
weave     Link encap:Ethernet  HWaddr f6:31:d7:4f:25:3e  
````
And here are the container interfaces
````
weave
eth0      Link encap:Ethernet  HWaddr 02:42:ac:11:02:8e  
          inet addr:172.17.2.142  Bcast:0.0.0.0  Mask:255.255.0.0
ethwe     Link encap:Ethernet  HWaddr 02:53:0d:f5:c7:8a  

weaveproxy
(same as host)

baz (aanand/docker-dnsutils)
eth0      Link encap:Ethernet  HWaddr 02:42:ac:11:02:90  
          inet addr:172.17.2.144  Bcast:0.0.0.0  Mask:255.255.0.0
ethwe     Link encap:Ethernet  HWaddr 2a:e7:54:ea:b4:96  
          inet addr:10.32.0.2  Bcast:0.0.0.0  Mask:255.240.0.0

foo (gliderlabs/alpine)
eth0      Link encap:Ethernet  HWaddr 02:42:ac:11:02:a7  
          inet addr:172.17.2.167  Bcast:0.0.0.0  Mask:255.255.0.0
ethwe     Link encap:Ethernet  HWaddr 9a:d4:b1:3a:4d:04  
          inet addr:10.32.0.1  Bcast:0.0.0.0  Mask:255.240.0.0
````
The packet capture shows a (small) number of ARPs:

- First we have 1a:65:93:46:8d:a1 (172.17.42.1) asks for 172.17.2.114 (unicast). This is followed by a 'DELETE /peer' to that IP, so it looks like the asked for IP was the *old* weave container, and the source was the temporary weaveexec container (we issue `DELETE /peer` only as part of `weave reset`, which in this case is issued by the test script). But that shouldn't have its own MAC address.

- Then we have 32:fc:9e:71:ac:8c (172.17.42.1) asks for 172.17.2.142 (broadcast). This seems to trigger a SYN & RST on three port 6784 (i.e. weave API) connections. The MAC is the from the first veth on the host, which is the docker0 link end of the 'weave' container. This is weird. Why would weave be ARPing for its own IP?

- Next we have 06:c6:e2:10:9a:be (172.17.42.1) asks for 172.17.2.142 (broadcast). This is followed by POSTs to the weave API. The MAC is from the 2nd veth on the host, which is the docker0 link end of the 'baz' container. This is odd, since that container should not be talking to the weave API.

- Next we have 06:c6:e2:10:9a:be (172.17.42.1) asks for 172.17.2.144 (unicast) This looks like baz ARPing for its own IP.

- Finally we have 02:c8:86:c4:35:f5 (172.17.42.1) asks for 172.17.2.142 (broadcast). This appears to be he proxy container since it subsequently POSTs to the weave API. But note that the MAC does not show up in our list, and that the proxy container runs in the host network ns.

The DNS query that goes astray comes shortly after that last ARP. It has a source of 02:42:ac:11:02:90 (172.17.2.144.144) - the baz 'eth0' and a destination of 02:c8:86:c4:35:f5 (172.17.42.1). Unlike any of the DNS requests before, which have a destination MAC of 06:c6:e2:10:9a:be.

**Comment by rade** [comment]
The root cause turns out to be the same as #1171. If I explicitly set the `docker0` MAC with `ip link set dev docker0 address EE:EE:EE:EE:EE:EE` then the test runs to completion without any stalls.

**Comment by awh** [comment]
Closed by #1229.

**Comment by rade** [comment]
Closed by #1231.
